### ISSUE TITLE
[1] Reduce Helmfile concurrency for improved predictability

### ISSUE THREAD

**Issue by cristiklein** [elastisys/compliantkubernetes-apps#402]
**Describe the bug**

This is a continuation of #391 and #392. 

We have quite a few webhook/CRDs-related failures on installing `compliantkubernetes-apps` that are due to too much concurrency. Steps that "change Kubernetes" by installing webhooks should not run in parallel to anything else.

The bootstrap steps feels like an out-of-helmfile solution to serialization. The Helmfile documentation suggests (and @viktor-f confirms) that separate Helmfile-s are serialized.

Let's come up with a proper serialization to improve reliability of applying `compliantkubernetes-apps`. I suggest the following structure:

```
00-namespaces.yaml => installs and labels namespaces (previously the bootstrap step)
01-gatekeeper.yaml => installs Gatekeeper
02-prometheus-operator.yaml
10-cert-manager.yaml
11-nginx-ingress.yaml
50-everything-else.yaml
```

If all works well, the CRDs-bootstrap step should no longer be needed.

**Additional context**

Might be related to "dev" pains and various complains on webhook failures that require re-running the apply step.

**Comment by OlleLarsson** [comment]
We should probably discuss the CRDs.
As I noted, and on which @lentzi90 commented - there is no lifecycle management of CRDs with helm.
How would the crd upgrade process look like? How would it differ from how we manage crds right now?

**Comment by cristiklein** [comment]
My understanding [from here](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you) is that we would let Helm manage CRDs for us. CRDs are supposed to be seen as "immutable" API, i.e., you never change an existing CRD you only add a new one with a newer version.

Makes sense?

**Comment by OlleLarsson** [comment]
Not really.
How would we upgrade the the prometheus CRD from 0.45 to 046?
https://github.com/prometheus-community/helm-charts/commit/85ae0d4c55809892e225389177d7d1dac02c3927#diff-23aa39170d284a5ef573c9e8f33725d41bea0c04b65157468d35dcfea430a2f9

**Comment by lentzi90** [comment]
Some clarification on what helm cannot do (from your link @cristiklein ):
> There is no support at this time for upgrading or deleting CRDs using Helm. This was an explicit decision after much community discussion due to the danger for unintentional data loss. Furthermore, there is currently no community consensus around how to handle CRDs and their lifecycle. As this evolves, Helm will add support for those use cases.

So yes, helm can *install* the CRDs the first time you install the chart, but after that it will simply ignore them. When you upgrade the chart and it brings a new version of the CRDs, you would have to apply them manually "in some way".

**Comment by cristiklein** [comment]
Okay, so first, thanks @OlleLarsson to bringing to my attention that we are using projects that "break" CRDs compatibility.

I'm a bit worried. :smile: Currently, we seem to simply `kubectl apply -f crds`, hoping that it will work:tm:. However, if there is no consensus in the community on how to deal with upgrading CRDs, we risk exposing ourselves to uncertainties and future issues. I mean, it kind of worked so far, but would we know if the new Prometheus operator likes to have its CRDs are suddenly changed? What about OPA/Gatekeeper? Velero?

At this point, I am tempted to suggest:
1. Let's identify which of our releases install breaking CRDs and put them mentally in a "special" box.
2. Let's identify which of our release depend on said CRDs and put them mentally in a "special" box.
3. Let's make sure we don't promise these CRDs as API to anyone.
4. Can we simply upgrade those releases by nuke-redeploy?

Rough analysis:
- cert-manager is rather stable, it should not bite us.
- prometheus-operator is rather unstable, but it should be possible to nuke-redeploy with only minor loss to monitoring data.
- velero should be nuke-redeploy-able.
- dex is stateless, so should be nuke-redeploy-able.

**Comment by lentzi90** [comment]
I think we need to be careful with all CRDs. Sure, some projects, like cert-manager, handle these things in a very good way with conversion hooks in order to support multiple simultaneous versions at the same time. But any changes to the CRDs is still a big deal. The thing that makes this even trickier is that we expose some of these CRDs to customers, which means that nuking and starting over would *delete any CustomResources that they have created* (e.g. Certificates or PrometheusRules).

Prometheus and Cert-manager are the ones I know we expose to customers directly (i.e. they can create CustomResources related to these operators). I'm not sure about Velero and Gatekeeper/OPA though.

**Comment by viktor-f** [comment]
> Prometheus and Cert-manager are the ones I know we expose to customers directly (i.e. they can create CustomResources related to these operators). I'm not sure about Velero and Gatekeeper/OPA though.

I users should not be able to create any of the velero/gatekeeper CRDs. Though they should perhaps be able to see some of the gatekeeper things, in order to see info about the audits.

**Comment by cristiklein** [comment]
Alrighty, so I looked more thoroughly at how upstream projects deal with CRDs, and here are my findings:

### cert-manager

* Installation: The cert-manager Helm Chart includes the [`installCRDs`](https://github.com/jetstack/cert-manager/blob/master/deploy/charts/cert-manager/values.yaml#L42) value -- by default it is set to `false`. If set to `true`, then CRDs are automatically installed when installing cert-manager, albeit not using the CRDs mechanism provided by Helm.
* Upgrade: CRDs are supposed to be [upgraded manually](https://cert-manager.io/docs/installation/upgrading/#upgrading-with-helm).

### dex

Unless I miss something, newer Dex installs [no CRDs](https://github.com/dexidp/helm-charts/tree/master/charts/dex). See #385.

### gatekeeper

* Installation: Gatekeeper installs CRDs using the [mechanism provided by Helm](https://github.com/open-policy-agent/gatekeeper/tree/master/charts/gatekeeper/crds).
* Upgrade: Gatekeeper wants you to either uninstall-install or run a [helm_migrate.sh](https://github.com/open-policy-agent/gatekeeper/tree/master/charts/gatekeeper#upgrade-chart).

### Prometheus (kube-prometheus-stack)

* Installation: kube-prometheus-stack installs CRDs using [standard Helm mechanism](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack/crds).
* Upgrade: kube-prometheus-stack expects you to [run manual upgrade steps](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack#from-14x-to-15x).

### Velero

* Installation: Velero install CRDs using [standard Helm mechanism](https://github.com/vmware-tanzu/helm-charts/tree/main/charts/velero/crds).
* Upgrade: Velero includes [magic to upgrade CRDs](https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/templates/upgrade-crds.yaml).

## Summary

All projects encourage installing CRDs as part of standard `helm install`.

Most projects encourage following manual migration steps to handle CRDs. Some projects handle CRD upgrades.

I find the current bootstrap process error-prone, un-intuitive and dangerous. Error-prone because we are currently installing the wrong CRDs, e.g., we install [Velero CRDs v2.8.2](https://github.com/elastisys/compliantkubernetes-apps/blob/main/bootstrap/crds/get-vendor-crds.sh#L18), but use [Velero v2.15.0](https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/01-applications.yaml#L117). Un-intuitive, because I would not expect to see CRDs being applied every time I configure CK8s. Dangerous, because upgrading CRDs is seen as special by some projects (as Lennart said) and we should keep it special.

## Proposal

* Remove all CRDs from the bootstrap step.
* Installation: Rely on upstream's approach to install CRDs.
* Upgrade: Propagate upstream migration steps in CK8s migration steps in each release notes.

Shall I draft an ADR to be discussed at the next architecture meeting? Happy to get more input on my proposal, both in this issue and during the arch meeting.

**Comment by lentzi90** [comment]
Great investigation @cristiklein !
I agree with most of it, but was a bit surprised by this:
> I would not expect to see CRDs being applied every time I configure CK8s
> ...
> Remove all CRDs from the bootstrap step.

To me the bootstrap step is there exactly because it provides a way to treat some things differently, and we could for example make it so that it is not run as part of a normal apply.

With Dex I think we can choose freely what storage to use. The default in the helm chart seems to be memory, which I'm not very happy with. Maybe it would be OK if we can guarantee that the dex Pod is very rarely restarted but it feels risky. Why they have set memory to default I don't know, maybe just to simplify deployment? It seems like dex will now create the CRDs itself if configured to use them and this is why it [needs RBAC to handle CRDs](https://github.com/dexidp/helm-charts/blob/master/charts/dex/templates/rbac.yaml).

**Comment by cristiklein** [comment]
> Great investigation @cristiklein !

Thanks for the feedback @lentzi90 .

> I agree with most of it, but was a bit surprised by this:
> 
> > I would not expect to see CRDs being applied every time I configure CK8s
> > ...
> > Remove all CRDs from the bootstrap step.
> 
> To me the bootstrap step is there exactly because it provides a way to treat some things differently, and we could for example make it so that it is not run as part of a normal apply.

To me, not running bootstrap as part of a normal apply would be one step in the right direction. However, looking at how upstream projects handle CRDs, I feel that aligning with the practices encouraged in those projects is more POLA-ish to me, and would reduce maintenance effort on our side. What are the benefits of keeping an out-of-chart copy of CRDs and applying them separately the first time, when `helm install` already does that for us?

Hence, I perceive the bootstrap step a legacy of pre-Helm-3 times which is no longer needed.

> 
> With Dex I think we can choose freely what storage to use. The default in the helm chart seems to be memory, which I'm not very happy with. Maybe it would be OK if we can guarantee that the dex Pod is very rarely restarted but it feels risky. Why they have set memory to default I don't know, maybe just to simplify deployment? It seems like dex will now create the CRDs itself if configured to use them and this is why it [needs RBAC to handle CRDs](https://github.com/dexidp/helm-charts/blob/master/charts/dex/templates/rbac.yaml).

This surprises me. Why do we persist Dex? To my understanding, you only need Dex persistence to complete the upstream OAuth2 flow. However, between authentication requests, Dex can be nuke-redeployed and nobody would notice. Unless someone complains, I perceive persistence with Dex to bring more complications than benefits, and propose disabling it.

**Comment by OlleLarsson** [comment]
> To me, not running bootstrap as part of a normal apply would be one step in the right direction.

We already have this - https://github.com/elastisys/compliantkubernetes-apps/blob/main/bin/ck8s#L15

**Comment by lentzi90** [comment]
> Why do we persist Dex?

I don't know the details, but this was the default until recently. People would at least be forced to login again if Dex was nuked, right? If it truly is no big deal to loose the state then I don't get why they have so many [storage options](https://dexidp.io/docs/storage/) (etcd, kubernetes CRDs, various SQL databases).

**Comment by OlleLarsson** [comment]
It appears as if the /healthz is responding before the webhook handeling components have started.
This means that condition for helm --wait is met and other releases can be installed, which of course is a problem if the webhooks aren't being served properly and we try to install new applications.
> Error: Internal error occurred: failed calling webhook "check-ignore-label.gatekeeper.sh": Post "https://gatekeeper-webhook-service.gatekeeper-system.svc:443/v1/admitlabel?timeout=3s": dial tcp 10.233.36.213:443: connect: connection refused

Procedure for gatekeeper:
- Option 1: Change liveness probe for gatekeeper pods to endpoint which server the (admitlabel) hook. Not there probably is no way of configuring this through [helm values](https://github.com/open-policy-agent/gatekeeper/blob/v3.3.0/charts/gatekeeper/values.yaml) currently. We would probably also have to make sure that there is nothing being installed concurrently with the chart.
- Option 2 (hack): Wait to install other releases until we know that the webhook is being served.
  This could be done with a for/while loop in the `scripts/deploy-wc.sh` script where we curl the webhook and with for a non connection refused response from the call.

**Comment by cristiklein** [comment]
Just making a note that this is still an issue with v0.15.0 when applied on the WC.
